### PR TITLE
Adds a new element, Self-Knockback.

### PR DESCRIPTION
**Pull request by KomradeSpectre**
## About The Pull Request
Adds a new element, Self Knockback, currently available for items and projectiles. This element will throw a user back a set amount of tiles proportional to the force of an item or damage of a projectile, min 1, max 5 tile(s) on things that contain this behavior. 

**I have a bespoke implementation of this ready locally, but will wait to see if the bespoke pr is merged to commit it. This will allow people to skip my damage-to-throw-distance calc with their own throw distance value.**

Edit: Forgot to mention I had added a new signal on projectile/fire() to accommodate the projectile behavior, but can be used by other things. The signal is sent as the projectile fires.

## Why It's Good For The Game
It's the inverse of our current knockback, and could be useful for things like recoil throwing users back on guns dependent on the projectile fired from them (though it's also available for items too). I'll be using it in my upcoming Gunmods system, so this is part of atomizing my stuff to prepare for that PR soon.

## Changelog
:cl: TheMythicGhost
add: Adds a self knockback (recoil) element that can be applied currently to items and projectiles.
/:cl:

### CODE REVIEW

**Comment by nemvar** [comment]
This feels like something that could benefit from https://github.com/tgstation/tgstation/pull/47807.

**Comment by KomradeSpectre** [comment]
> This feels like something that could benefit from #47807.

I actually structured this according to advice from @ninjanomnom himself.

File: code/datums/elements/selfknockback.dm [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	if(ismovableatom(attacktarget)) //This check prevents us from launching ourselves off floors or walls.
+		if(proximity_flag || (get_dist(attacktarget, usertarget) <= I.reach))
+			var/target_angle = Get_Angle(attacktarget, usertarget)
+			var/knockback_force = CLAMP(CEILING(I.force / 10, 1), 1, 5)
+			var/move_target = get_ranged_target_turf(usertarget, angle2dir(target_angle), knockback_force)
+			usertarget.throw_at(move_target, knockback_force, knockback_force)
+			usertarget.visible_message("<span class='warning'>[usertarget] gets thrown back by the force of \the [I] impacting \the [attacktarget]!</span>", "<span class='warning'>The force of \the [I] impacting \the [attacktarget] sends you flying!</span>")
+
+/datum/element/selfknockback/proc/Projectile_SelfKnockback(obj/projectile/P)
+	var/atom/movable/knockback_target = P.firer
```
Comment (nemvar) [inline code/datums/elements/selfknockback.dm]: Not every projectile has a firer. Also feels kinda odd to attach this to projectiles in the first place.

**Comment by nemvar** [comment]
What I meant is that knockback is based on force. This means that you can't, for example, have a lasertag gun that has knockback. That's why it could benefit from bespoke elements

**Comment by KomradeSpectre** [comment]
> 
> 
> What I meant is that knockback is based on force. This means that you can't, for example, have a lasertag gun that has knockback. That's why it could benefit from bespoke elements

Did you read the changed files? Projectiles are based on damage, items are based on force.

**Comment by nemvar** [comment]
The point still stands that lasertag shots can't cause knockback.

**Comment by KomradeSpectre** [comment]
> 
> 
> The point still stands that lasertag shots can't cause knockback.

Projectile
`var/knockback_force = CLAMP(CEILING((P.damage / 10), 1), 1, 5)`
Item
`var/knockback_force = CLAMP(CEILING(I.force / 10, 1), 1, 5)`

**Comment by KomradeSpectre** [comment]
Fixed a potential runtime error before it can happen thanks to @nemvar , after discussion I will look into making this bespoke after/if ninjanomnom's bespoke PR is merged.

**Comment by GuyonBroadway** [comment]
As a suggestion perhaps consider not directly tying knockback to damage so you can have a 0 damage object with massive recoil for shits and/or giggles.

**Comment by KomradeSpectre** [comment]
> 
> 
> As a suggestion perhaps consider not directly tying knockback to damage so you can have a 0 damage object with massive recoil for shits and/or giggles.

The issue with that is force isn't the universal damage, neither is damage. It's typed because the implementations are different.

**Comment by GuyonBroadway** [comment]
Sorry I meant force.

Indeed it'd be funny to say, get some item out the arcade that sends you flying back 10 tiles if you throw it even if said item was otherwise harmless.

**Comment by KomradeSpectre** [comment]
That should be possible if I pursue using the Element Bespoke if it gets merged.

File: code/datums/elements/selfknockback.dm [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/datum/element/selfknockback
+	element_flags = ELEMENT_DETACH
```
Comment (ninjanomnom) [inline code/datums/elements/selfknockback.dm]: You don't need this flag, signals get cleaned up automaticaly in qdel. Keep the code in Detach though

File: code/datums/elements/selfknockback.dm [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+		RegisterSignal(target, COMSIG_PROJECTILE_FIRE, .proc/Projectile_SelfKnockback)
+	else	
+		return ELEMENT_INCOMPATIBLE
+	
+/datum/element/selfknockback/Detach(datum/source, force)
+	. = ..()
+	UnregisterSignal(source, list(COMSIG_ITEM_AFTERATTACK, COMSIG_PROJECTILE_FIRE))
+
+/datum/element/selfknockback/proc/Item_SelfKnockback(obj/item/I, atom/attacktarget, mob/usertarget, proximity_flag)
+	if(ismovableatom(attacktarget)) //This check prevents us from launching ourselves off floors or walls.
```
Comment (ninjanomnom) [inline code/datums/elements/selfknockback.dm]: Why?

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by ATH1909** [comment]
Could you use this component for the effect of firing an improvised pneumatic cannon at the maximum pressure setting?

**Comment by KomradeSpectre** [comment]
> 
> 
> Could you use this component for the effect of firing an improvised pneumatic cannon at the maximum pressure setting?

The two implementations I included here are for item listening for COMSIG_ITEM_AFTERATTACK, and projectile listening for COMSIG_PROJECTILE_FIRE. So, the element is currently compatible with attacking someone/something with an item, or firing a bullet on compatible things with this attached.

Additional note/update, I've locally made this Bespoke ready, and people can input a custom value during the AddElement process to skip my damage calc and input their own tile distance to throw should they want to.

**Comment by AnonymousNow** [comment]
Consider eliminating when using magboots, except in instances where the knockback exceeds a large threshold, at which point it just rips your body out of the boots without your feet.

**Comment by ATH1909** [comment]
> > Could you use this component for the effect of firing an improvised pneumatic cannon at the maximum pressure setting?
> 
> The two implementations I included here are for item listening for COMSIG_ITEM_AFTERATTACK, and projectile listening for COMSIG_PROJECTILE_FIRE. So, the element is currently compatible with attacking someone/something with an item, or firing a bullet on compatible things with this attached.
> 
> Additional note/update, I've locally made this Bespoke ready, and people can input a custom value during the AddElement process to skip my damage calc and input their own tile distance to throw should they want to.

No, I meant that maximum pressure improvised pneumatic cannons already do something similar to this when fired, and that you should make them use this system instead.

**Comment by KomradeSpectre** [comment]
> No, I meant that maximum pressure improvised pneumatic cannons already do something similar to this when fired, and that you should make them use this system instead.

>Consider eliminating when using magboots, except in instances where the knockback exceeds a large threshold, at which point it just rips your body out of the boots without your feet.

When and if this gets merged after I commit the bespoke thing (when and if the parent PR for THAT gets merged) then by all means feel free to PR something that adds this behavior where you want it.

**Comment by GuyonBroadway** [comment]
>Consider eliminating when using magboots, except in instances where the knockback exceeds a large threshold, at which point it just rips your body out of the boots without your feet.

Honestly imagine if you hit some hard enough you punted them out from under thier own hat/out thier own boots/both

**Comment by ninjanomnom** [comment]
Bespoke elements are merged if you want to add the knockback force argument now

**Comment by KomradeSpectre** [comment]
Will fix the newline issue after I test this, really quickly.

File: code/datums/elements/selfknockback.dm [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	override_throw_val = throw_amount
+	override_speed_val = speed_amount
+	
+/datum/element/selfknockback/Detach(datum/source, force)
+	. = ..()
+	UnregisterSignal(source, list(COMSIG_ITEM_AFTERATTACK, COMSIG_PROJECTILE_FIRE))
+
+/datum/element/selfknockback/proc/Get_Knockback_Force(default_force)
+	if(override_throw_val > 0)
+		return round(override_throw_val, 1)
```
Comment (ninjanomnom) [inline code/datums/elements/selfknockback.dm]: You don't need to round, if someone gives a decimal then it's their own fault

File: code/datums/elements/selfknockback.dm [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	else
+		return default_force
+
+/datum/element/selfknockback/proc/Get_Knockback_Speed(default_speed)
+	if(override_speed_val > 0)
+		return round(override_speed_val, 1)
```
Comment (ninjanomnom) [inline code/datums/elements/selfknockback.dm]: Ditto

File: code/datums/elements/selfknockback.dm [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	
+	override_throw_val = throw_amount
+	override_speed_val = speed_amount
+	
+/datum/element/selfknockback/Detach(datum/source, force)
+	. = ..()
+	UnregisterSignal(source, list(COMSIG_ITEM_AFTERATTACK, COMSIG_PROJECTILE_FIRE))
+
+/datum/element/selfknockback/proc/Get_Knockback_Force(default_force)
+	if(override_throw_val > 0)
```
Comment (ninjanomnom) [inline code/datums/elements/selfknockback.dm]:
```suggestion
	if(override_throw_val)
```

File: code/datums/elements/selfknockback.dm [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+		return round(override_throw_val, 1)
+	else
+		return default_force
+
+/datum/element/selfknockback/proc/Get_Knockback_Speed(default_speed)
+	if(override_speed_val > 0)
```
Comment (ninjanomnom) [inline code/datums/elements/selfknockback.dm]:
```suggestion
	if(override_speed_val)
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Review by ninjanomnom — APPROVED** [review]

**Pull request merged**